lnrpc/routerrpc/router_backend.go

```diff
@@ -844,35 +844,72 @@ func UnmarshalMPP(reqMPP *lnrpc.MPPRecord) (*record.MPP, error) {
 func (r *RouterBackend) MarshalHTLCAttempt(
 	htlc channeldb.HTLCAttempt) (*lnrpc.HTLCAttempt, error) {
 
-	var (
-		status      lnrpc.HTLCAttempt_HTLCStatus
-		resolveTime int64
-	)
+	route, err := r.MarshallRoute(&htlc.Route)
+	if err != nil {
+		return nil, err
+	}
+
+	rpcAttempt := &lnrpc.HTLCAttempt{
+		AttemptTimeNs: MarshalTimeNano(htlc.AttemptTime),
+		Route:         route,
+	}
 
 	switch {
 	case htlc.Settle != nil:
-		status = lnrpc.HTLCAttempt_SUCCEEDED
-		resolveTime = MarshalTimeNano(htlc.Settle.SettleTime)
+		rpcAttempt.Status = lnrpc.HTLCAttempt_SUCCEEDED
+		rpcAttempt.ResolveTimeNs = MarshalTimeNano(
+			htlc.Settle.SettleTime,
+		)
 
 	case htlc.Failure != nil:
-		status = lnrpc.HTLCAttempt_FAILED
-		resolveTime = MarshalTimeNano(htlc.Failure.FailTime)
+		rpcAttempt.Status = lnrpc.HTLCAttempt_FAILED
+		rpcAttempt.ResolveTimeNs = MarshalTimeNano(
+			htlc.Failure.FailTime,
+		)
 
+		var err error
+		rpcAttempt.Failure, err = marshallHtlcFailure(htlc.Failure)
+		if err != nil {
+			return nil, err
+		}
 	default:
-		status = lnrpc.HTLCAttempt_IN_FLIGHT
+		rpcAttempt.Status = lnrpc.HTLCAttempt_IN_FLIGHT
 	}
 
-	route, err := r.MarshallRoute(&htlc.Route)
-	if err != nil {
-		return nil, err
+	return rpcAttempt, nil
+}
+
+// marshallHtlcFailure marshalls htlc fail info from the database to its rpc
+// representation.
+func marshallHtlcFailure(failure *channeldb.HTLCFailInfo) (*lnrpc.Failure,
+	error) {
+
+	rpcFailure := &lnrpc.Failure{
+		FailureSourceIndex: failure.FailureSourceIndex,
 	}
 
-	return &lnrpc.HTLCAttempt{
-		Status:        status,
-		Route:         route,
-		AttemptTimeNs: MarshalTimeNano(htlc.AttemptTime),
-		ResolveTimeNs: resolveTime,
-	}, nil
+	switch failure.Reason {
+
+	case channeldb.HTLCFailUnknown:
+		rpcFailure.Code = lnrpc.Failure_UNKNOWN_FAILURE
+
+	case channeldb.HTLCFailUnreadable:
+		rpcFailure.Code = lnrpc.Failure_UNREADABLE_FAILURE
+
+	case channeldb.HTLCFailInternal:
+		rpcFailure.Code = lnrpc.Failure_INTERNAL_FAILURE
+
+	case channeldb.HTLCFailMessage:
+		err := marshallWireError(failure.Message, rpcFailure)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, errors.New("unknown htlc failure reason")
+	}
+
+	return rpcFailure, nil
```

Comment:
There're a few fields that we appear to not properly be setting (might as well give all the information we have): 
```
                    "failure": {
                        "code": "TEMPORARY_CHANNEL_FAILURE",
                        "channel_update": {
                            "signature": "2bc91eb05e6fe9e286e0e09372efdaafec4a62d224bdb7db7d23ea431722ac801ca39a9dd6827fe6d8913e806a1e6ed43ec7cae13a4e3bb8ffb40efbec6fdc11",
                            "chain_hash": "43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000",
                            "chan_id": "1832711061194407937",
                            "timestamp": 1583766633,
                            "message_flags": 1,
                            "channel_flags": 1,
                            "time_lock_delta": 15,
                            "htlc_minimum_msat": "1000",
                            "base_fee": 0,
                            "fee_rate": 0,
                            "htlc_maximum_msat": "16609154000",
                            "extra_opaque_data": null
                        },
                        "htlc_msat": "0",
                        "onion_sha_256": null,
                        "cltv_expiry": 0,
                        "flags": 0,
                        "failure_source_index": 1,
                        "height": 0
```